### ISSUE TITLE
GCC warns when compiling the JIT: too small to hold all values of regNumber

### ISSUE THREAD

**Issue by am11** [dotnet/runtime#33541]
e.g. around 78% of this build https://dev.azure.com/dnceng/public/_build/results?buildId=557918&view=logs&jobId=e5f41d17-fae4-5d3d-f998-568159b1c41c&j=e5f41d17-fae4-5d3d-f998-568159b1c41c&t=86eaea2a-2e8c-50dd-cc99-c57aff36e5ad, there are a lots of such (non-fatal) warnings:

```
2020-03-13T00:38:03.8562716Z [ 78%] Built target utilcode_dac
2020-03-13T00:38:04.4983399Z Scanning dependencies of target clrjit
2020-03-13T00:38:04.6247371Z [ 78%] Building CXX object src/jit/standalone/CMakeFiles/clrjit.dir/__/alloc.cpp.o
2020-03-13T00:38:05.3361944Z In file included from /__w/1/s/src/coreclr/src/jit/codegeninterface.h:27:0,
2020-03-13T00:38:05.3363711Z                  from /__w/1/s/src/coreclr/src/jit/compiler.h:52,
2020-03-13T00:38:05.3453068Z                  from /__w/1/s/src/coreclr/src/jit/jit.h:806,
2020-03-13T00:38:05.3454005Z                  from /__w/1/s/src/coreclr/src/jit/jitpch.h:19,
2020-03-13T00:38:05.3455061Z                  from /__w/1/s/src/coreclr/src/jit/alloc.cpp:5:
2020-03-13T00:38:05.3457223Z /__w/1/s/src/coreclr/src/jit/emit.h:522:51: warning: 'emitter::emitAddrMode::amBaseReg' is too small to hold all values of 'regNumber {aka enum _regNumber_enum}'
2020-03-13T00:38:05.3458916Z          regNumber       amBaseReg : REGNUM_BITS + 1;
2020-03-13T00:38:05.3459845Z                                                    ^
2020-03-13T00:38:05.3461568Z /__w/1/s/src/coreclr/src/jit/emit.h:523:51: warning: 'emitter::emitAddrMode::amIndxReg' is too small to hold all values of 'regNumber {aka enum _regNumber_enum}'
2020-03-13T00:38:05.3462801Z          regNumber       amIndxReg : REGNUM_BITS + 1;
2020-03-13T00:38:05.3463827Z                                                    ^
2020-03-13T00:38:05.3465436Z /__w/1/s/src/coreclr/src/jit/emit.h:524:35: warning: 'emitter::emitAddrMode::amScale' is too small to hold all values of 'enum emitter::opSize'
2020-03-13T00:38:05.3466617Z          emitter::opSize amScale : 2;
2020-03-13T00:38:05.3467320Z                                    ^
2020-03-13T00:38:05.3468033Z In file included from /__w/1/s/src/coreclr/src/jit/jit.h:707:0,
2020-03-13T00:38:05.3468843Z                  from /__w/1/s/src/coreclr/src/jit/jitpch.h:19,
2020-03-13T00:38:05.3469644Z                  from /__w/1/s/src/coreclr/src/jit/alloc.cpp:5:
2020-03-13T00:38:05.3471439Z /__w/1/s/src/coreclr/src/jit/target.h:557:36: warning: 'emitter::instrDesc::idAddrUnion::<unnamed struct>::_idReg3' is too small to hold all values of 'regNumber {aka enum _regNumber_enum}'
2020-03-13T00:38:05.3472852Z    #define REGNUM_BITS              6       // number of bits in a REG_*
2020-03-13T00:38:05.3473658Z                                     ^
2020-03-13T00:38:05.3475339Z /__w/1/s/src/coreclr/src/jit/emit.h:847:37: note: in expansion of macro 'REGNUM_BITS'
2020-03-13T00:38:05.3476435Z                  regNumber _idReg3 : REGNUM_BITS;
2020-03-13T00:38:05.3477214Z                                      ^~~~~~~~~~~
2020-03-13T00:38:05.3478962Z /__w/1/s/src/coreclr/src/jit/target.h:557:36: warning: 'emitter::instrDesc::idAddrUnion::<unnamed struct>::_idReg4' is too small to hold all values of 'regNumber {aka enum _regNumber_enum}'
2020-03-13T00:38:05.3480334Z    #define REGNUM_BITS              6       // number of bits in a REG_*
2020-03-13T00:38:05.3481124Z                                     ^
2020-03-13T00:38:05.3482349Z /__w/1/s/src/coreclr/src/jit/emit.h:848:37: note: in expansion of macro 'REGNUM_BITS'
2020-03-13T00:38:05.3483322Z                  regNumber _idReg4 : REGNUM_BITS;
2020-03-13T00:38:05.3484102Z                                      ^~~~~~~~~~~
2020-03-13T00:38:05.3484918Z In file included from /__w/1/s/src/coreclr/src/jit/codegeninterface.h:27:0,
2020-03-13T00:38:05.3486183Z                  from /__w/1/s/src/coreclr/src/jit/compiler.h:52,
2020-03-13T00:38:05.3487168Z                  from /__w/1/s/src/coreclr/src/jit/jit.h:806,
2020-03-13T00:38:05.3488054Z                  from /__w/1/s/src/coreclr/src/jit/jitpch.h:19,
2020-03-13T00:38:05.3495192Z                  from /__w/1/s/src/coreclr/src/jit/alloc.cpp:5:
2020-03-13T00:38:05.3498610Z /__w/1/s/src/coreclr/src/jit/emit.h:570:30: warning: 'emitter::instrDesc::_idIns' is too small to hold all values of 'enum instruction'
2020-03-13T00:38:05.3499774Z          instruction _idIns : 10;
```

Is this something actionable?

category:implementation
theme:build
skill-level:beginner
cost:small

**Comment by janvorli** [comment]
It looks like the regNumber enum has two extra members that are never stored in the fields the compiler complains about, see:
```c++
enum _regNumber_enum : unsigned
{
#define REGDEF(name, rnum, mask, sname) REG_##name = rnum,
#define REGALIAS(alias, realname) REG_##alias = REG_##realname,
#include "register.h"

    REG_COUNT,
    REG_NA           = REG_COUNT,
    ACTUAL_REG_COUNT = REG_COUNT - 1 // everything but REG_STK (only real regs)
};
```

**Comment by am11** [comment]
I have found a related bug: https://gcc.gnu.org/bugzilla/show_bug.cgi?id=61414, which seems to be fixed last month in 8.4+ and 9.3+, by introducing (only?) a warning suppression.

The problem is that when the enum with fixed underlying type is used as a bit fields in a struct (or class), and has any valid value (within limits) different width than the underlying type, we get this unspresseble warning (hardcoded in g++). Minimal repro that demonstrates the issue with GCC when it compiles JIT code: https://godbolt.org/z/YZnRvq.

**Comment by am11** [comment]
Update: the false-positive bug (https://godbolt.org/z/YZnRvq) was fixed in gcc 9.3+, 8.4+ and 10.0+. The GCC CI leg in master branch was today upgraded from gcc 7.3.1 to 9.3.1.

The remaining bitfield warnings are real, but not critical. However, they are on the edge of being critical, it seems.

Actually both GCC and Clang complain about e.g. https://github.com/dotnet/runtime/blob/0e951b138982495f7d37c6be42f6a6030ad99a87/src/coreclr/jit/emit.h#L516
because the largest value is 6, which requires 3 bits https://github.com/dotnet/runtime/blob/0e951b138982495f7d37c6be42f6a6030ad99a87/src/coreclr/jit/emit.h#L439

GCC: `warning: 'emitter::emitAddrMode::amScale' is too small to hold all values of 'enum emitter::opSize'`
Clang with `-Weverything`:   `warning: bit-field 'amScale' is not wide enough to store all enumerators of 'opSize' [-Wbitfield-enum-conversion]`

They are currently not critical because we are not _assigning_ the values which exceed the range. Otherwise clang without `-Weverything` issues `-Wbitfield-constant-conversion` and gcc produces `-Woverflow`. On runtime, the program compiled by either assigns garbage or zero for values out of bitfield range; which would make it a critical bug (but I think we have some assertions in emit.h to prevent them from happening).
